### PR TITLE
PARQUET-835: Read Arrow columns in parallel with thread pool

### PR DESCRIPTION
**Pull request by wesm**
Also implements PARQUET-836, but need to add a unit test for that

### CODE REVIEW

File: src/parquet/arrow/arrow-reader-writer-test.cc [hunk 1]
```diff
@@ -184,6 +185,23 @@ using ParquetDataType = DataType<test_traits<T>::parquet_enum>;
...
+
+  std::shared_ptr<Buffer> buffer = sink->GetBuffer();
+  std::unique_ptr<FileReader> reader;
+  ASSERT_OK_NO_THROW(
+      OpenFile(std::make_shared<BufferReader>(buffer), ::arrow::default_memory_pool(),
+          ::parquet::default_reader_properties(), nullptr, &reader));
+
+  reader->set_num_threads(num_threads);
+  ASSERT_OK_NO_THROW(reader->ReadFlatTable(out));
+}
```
Comment (wesm) [inline src/parquet/arrow/arrow-reader-writer-test.cc]: @xhochy I'm wondering if this test module can be simplified using a slightly more generalized version of this function

**Comment by xhochy** [comment]
In general this seems fine but once we would add S3 support to Arrow IO, this would definitely need to be reworked. The I/O is hidden inside of this code and would need to be more explicit to be tunable for the underlying storage access cost model. 

Not saying that parallel I/O doesn't work with object stores (it actually can be beneficial) but it behaves differently than to an SSD where you can seek randomly.

**Comment by wesm** [comment]
We should look at Impala's S3 support for inspiration (they spent a long time designing / building it as I recall).

**Comment by xhochy** [comment]
The only thing I found in Impala's source code was that they actually redirect to HDFS (read the s3a implementation in Hadoop) in `be/src/runtime/hdfs-fs-cache.{h,cc}` and don't actually implement it in native code.

**Comment by wesm** [comment]
Thanks. I'll add a unit test for PARQUET-836 in a follow up patch

**Pull request closed without merging**